### PR TITLE
IXLSheetProtection refactoring

### PR DESCRIPTION
**Pull request by igitur**
Fixes https://stackoverflow.com/questions/52864198/closed-xml-protect-worksheet-dont-allow-edit-objects

Refactors `IXLSheetProtection` to make the meaning of the properties and methods more obvious.

### CODE REVIEW

**Comment by igitur** [comment]
@b0bi79 This is a breaking change that might affect ClosedXML.Report. This PR is scheduled for inclusion in ClosedXML v0.95.

**Comment by igitur** [comment]
Thanks. Good points. Maybe it's just easier to disable the `Protection` setter? By default it would be initialized, but with `IsEnabled = false`.

Also, while I'm working on this, maybe we should switch to a flag based implementation. What do you think?

**Comment by igitur** [comment]
> Yet another option is to implement the Protection setter for it to perform a shadow copying of the passed value. This again requires a Clone() method in IXLSheetProtection.

I'm currently implementing something similar to this. Can you see any issues with setting the destination spreadsheet's protection password to the original spreadsheet's protection password?

**Comment by Pankraty** [comment]
> Can you see any issues with setting the destination spreadsheet's protection password to the original spreadsheet's protection password? See the last commit in this PR.

I see no issues. The password hash is available anyway, so there is no security flaw.

**Comment by igitur** [comment]
@Pankraty  I've refactored this to use a flag enum and I've implement `ICloneable` on `IXLSheetProtection` as per your comments.

File: ClosedXML/Excel/XLWorkbook_Save.cs [hunk 1]
```diff
@@ -4976,23 +4976,28 @@ private static bool NumberFormatsAreEqual(NumberingFormat nf, XLNumberFormatValu
...
+                sheetProtection.InsertHyperlinks = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.InsertHyperlinks), true);
+                sheetProtection.DeleteColumns = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.DeleteColumns), true);
+                sheetProtection.DeleteRows = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.DeleteRows), true);
+                sheetProtection.Sort = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.Sort), true);
+                sheetProtection.AutoFilter = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.AutoFilter), true);
+                sheetProtection.PivotTables = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.PivotTables), true);
+                sheetProtection.Scenarios = OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(XLSheetProtectionElements.EditScenarios), true);
```
Comment (Pankraty) [inline ClosedXML/Excel/XLWorkbook_Save.cs]: To make the assignments less verbose we can define a local function 
```
Func<XLSheetProtectionElements, bool, BooleanValue> flagToBooleanValue = (flag, defaultValue) => OpenXmlHelper.GetBooleanValue(!protection.AllowedElements.HasFlag(flag), defaultValue);

sheetProtection.FormatCells = flagToBooleanValue(XLSheetProtectionElements.FormatCells, true);
...
```
Or maybe a nested method.

File: ClosedXML/Excel/IXLSheetProtection.cs [hunk 2]
```diff
@@ -1,46 +1,78 @@
...
+        InsertRows = 1 << 11,
+        PivotTables = 1 << 12,
+        SelectLockedCells = 1 << 13,
+        SelectUnlockedCells = 1 << 14,
+        Sort = 1 << 15,
+
+        SelectEverything = SelectLockedCells | SelectUnlockedCells,
+        DeleteEverything = DeleteColumns | DeleteRows,
+        FormatEverything = FormatCells | FormatColumns | FormatRows,
+        InsertEverything = InsertColumns | InsertHyperlinks | InsertRows,
```
[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML/Excel/IXLSheetProtection.cs]: Ah, it was supposed to be ordered. How did I miss that one!

File: ClosedXML/Excel/XLWorkbook_Save.cs (hunk 1 above)
Comment (igitur) [reply ClosedXML/Excel/XLWorkbook_Save.cs]: To be honest, `OpenXmlHelper.GetBooleanValue()` is used in many places, so I'm comfortable to use it directly here too. I don't think it is too verbose.

File: ClosedXML/Excel/IXLSheetProtection.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (igitur) [reply ClosedXML/Excel/IXLSheetProtection.cs]: Fixed

**Pull request merged**